### PR TITLE
Fix data race

### LINKED ISSUE #101097: staging/src/k8s.io/apiserver/pkg/server/filters Data race
#### What happened:

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101093/pull-kubernetes-unit/1382237738274656256

``` log
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters: TestErrConnKilledHTTP2 expand_less	0s
=== RUN   TestErrConnKilledHTTP2
==================
WARNING: DATA RACE
Write at 0x00000352f1a8 by goroutine 216:
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.captureStdErr.func2()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout_test.go:220 +0x5a
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.TestErrConnKilledHTTP2()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout_test.go:367 +0xe3f
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202

Previous read at 0x00000352f1a8 by goroutine 60:
  k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).output()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:919 +0xa64
  k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).printDepth()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:733 +0x28c
  k8s.io/kubernetes/vendor/k8s.io/klog/v2.ErrorDepth()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1451 +0x12d
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logError()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:114 +0x2f
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleError()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:108 +0xa1
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func2.1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:126 +0x125


```

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?
/kind bug
/kind flake

#### Which issue(s) this PR fixes:

Fixes #101097

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go [hunk 1]
```diff
@@ -217,11 +223,11 @@ func captureStdErr() (func() string, func(), error) {
 	}
 	klog.LogToStderr(true)
 	cleanUp := func() {
-		os.Stderr = stderr
+		safeSetStderr(stderr)
 		klog.LogToStderr(false)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:227]: The log in the linked issue showed that the race was between this test setting os.Stderr and klog calling to os.Stderr.Write.

So if we run `klog.LogToStderr(false)` first, before running `os.Stderr = stderr`, does that solve the data race? That's a cleaner fix.

Comment (wzshiming) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:219]: Thank you, seems feasible.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go [hunk 2]
```diff
@@ -215,12 +215,13 @@ func captureStdErr() (func() string, func(), error) {
 		<-readerClosedCh
 		return buf.String()
 	}
-	klog.LogToStderr(true)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:218]: I don't think we need to change this line.

I think we only need to change the order of `os.Stderr=stderr` and `klog.LogToStderr(false)` in the cleanup function.

Comment (wzshiming) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:218]:
``` diff
-	klog.LogToStderr(true)
	cleanUp := func() {
+		klog.LogToStderr(false)
		os.Stderr = stderr
-		klog.LogToStderr(false)
		stopReadingStdErr()
	}

+	klog.LogToStderr(false)
	os.Stderr = writer
	go func() {
		io.Copy(&buf, reader)
		readerClosedCh <- struct{}{}
		close(readerClosedCh)
	}()
	klog.LogToStderr(true)
```

I think `klog.LogToStderr(false)` must have before every `os.Stderr` modification.
There is also an `klog.LogToStderr(true)` at the bottom of this function to reenable Log

**Comment by wzshiming** [comment]
friendly ping @caesarxuchao

**Comment by wzshiming** [comment]
@caesarxuchao PTAL

**Comment by pacoxu** [comment]
This flake still exists recently. I opened #103246 which should be fixed by this PR as well.

**Comment by pacoxu** [comment]
> This flake still exists recently. I opened #103246 which should be fixed by this PR as well.

kindly ping @liggitt

**Pull request closed without merging**